src/cudamatrix/cu-kernels.cu

```diff
@@ -2629,19 +2629,109 @@ static void _diff_log_softmax(const MatrixDim in_deriv_dim,
   }
 }
 
-/// Add spatial regularization term to `out_deriv`.
+/// This CUDA kernel implements the idea 'spatial smoothing' as described in
+/// the paper https://arxiv.org/pdf/1610.05256v1.pdf
+///
+/// It adds spatial regularization term to `out_deriv`.
 /// The regularization term is calculated from `out_value`.
-/// Each row of `out_value` is first viewed as a 16*n grid,
-/// convoluted with the 3x3 kernel, scaled with `scale`
+/// Each row of `out_value` is first viewed as a n*16 or n*32 2D image,
+/// high-pass filtered by a 3x3 kernel, scaled with `scale`
 /// and then added to `out_deriv`.
 /// If regularization_sqsum is not NULL, the squared sum of the
 /// regularization term will be calculated and stored.
+///
+/// This CUDA kernel uses 2D grid. Each row of the grid handles one row of
+/// data matrices including `out_value` and `out_deriv`.
+/// Each thread block contains 256 threads. Each thread handles one element
+/// of the data matrices, thus we may need more than one block per grid row to
+/// cover the matrix row dim.
+/// The thread block also has 2D, with a row width of 16/32, matching the width
+/// of the image mentioned above.
 template<typename Real>
 __global__
 static void _add_spatial_regularization_deriv(
     const Real* out_value, const MatrixDim out_value_dim, Real* out_deriv,
-    const int out_deriv_stride, Real* regularization_sqsum,
+    const int out_deriv_stride, const Real scale, Real* regularization_sqsum,
     const int regularization_sqsum_stride) {
+  const unsigned int kBlockSize = 256;
+  const unsigned int kSectWidth = 16;
+  __shared__ union {
+    Real buf[kBlockSize/kSectWidth+2][kSectWidth];
+    Real sum[kBlockSize];
+  } smem;
+
+  const int i = blockIdx.y;
+  const int tid = threadIdx.y * blockDim.x + threadIdx.x;
+  const int j = blockIdx.x * kBlockSize + tid;
+
+
+  // Cache global mem data `out_value` to shared mem
+  if (threadIdx.y == 0) {
+    // Leading apron
```

Comment:
'apron' seems to be some obscure NVidia jargon.. please explain in simpler terms what is going on here.